kv/txn_coord_sender.go

```diff
@@ -407,20 +405,26 @@ func (tc *TxnCoordSender) sendBatch(batchArgs *proto.InternalBatchRequest, batch
 	// TODO(spencer): send calls in parallel.
 	batchReply.Txn = batchArgs.Txn
 	for i := range batchArgs.Requests {
-		// Initialize args header values where appropriate.
 		args := batchArgs.Requests[i].GetValue().(proto.Request)
 		call := client.Call{Args: args}
-		if args.Header().User == "" {
-			args.Header().User = batchArgs.User
+		// Disallow transaction, user and priority on individual calls, unless
+		// equal.
+		if args.Header().User != "" && args.Header().User != batchArgs.User {
+			batchReply.Header().SetGoError(util.Error("cannot have individual user on call in batch"))
+			return
 		}
-		if args.Header().UserPriority == nil {
-			args.Header().UserPriority = batchArgs.UserPriority
+		args.Header().User = batchArgs.User
+		if args.Header().UserPriority != nil && args.Header().GetUserPriority() != batchArgs.GetUserPriority() {
+			batchReply.Header().SetGoError(util.Error("cannot have individual user priority on call in batch"))
+			return
 		}
-		// Only update the individual calls' Txn from the batch if there isn't
-		// one already.
-		if args.Header().Txn == nil {
-			args.Header().Txn = batchArgs.Txn
+		args.Header().UserPriority = batchArgs.UserPriority
+		if args.Header().Txn != nil && args.Header().GetTxn() != batchArgs.GetTxn() {
```

Comment:
Even if it returned a `proto.Transaction` instead of a `*proto.Transaction`, there would be a problem, because the Transaction struct includes fields with pointer types. These will compare as equal only if they point to the same memory (which may be true in local-only tests, but will not be true once things have passed through the DistSender and been marshaled and unmarshaled).